### PR TITLE
Update to 8.8.0.121 and add CUDA 12 support.

### PR DESCRIPTION
**Pull request by bdice**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

This PR updates cuDNN to 8.8.0.121 with support for CUDA 11 and CUDA 12.

Fixes https://github.com/conda-forge/cudnn-feedstock/issues/54

### CODE REVIEW

**Comment by bdice** [comment]
@conda-forge-admin, please rerender

**Comment by bdice** [comment]
cc: @jakirkham @leofang @adibbley @robertmaynard for notice/review. I think CI should pass this round (still running).

**Comment by bdice** [comment]
(and thanks in advance to @kkraus14 as package maintainer for a review, as well) 🙏

File: recipe/meta.yaml [hunk 1]
```diff
@@ -45,36 +47,46 @@ requirements:
...
+    {% else %}
+    - cuda-cudart-dev
+    - cuda-nvrtc
+    - libcublas
+    {% endif %}
   run_constrained:
-    # Only GLIBC_2.17 or older symbols present
-    - __glibc >=2.17  # [linux]
+    - __glibc >=2.17                        # [linux and not (aarch64 and cuda_major == "12")]
+    - __glibc >=2.28                        # [linux and (aarch64 and cuda_major == "12")]
```
Comment (leofang) [inline recipe/meta.yaml:70]: I think this should be just
```suggestion
    - __glibc >=2.17                        # [linux and not aarch64]
    - __glibc >=2.28                        # [linux and aarch64]
```
(I think `linux` can be omitted too). I see the glibc 2.27 symbols in CUDA 11 binaries as well.

**Review by leofang — COMMENTED** [review]
Thanks, Bradley! LGTM except for one nit.

**Comment by leofang** [comment]
Q: Would it possible that we try moving ppc64le back to Travis? It is slightly stabler than before, so if it can complete aarch64 in less than 15 mins, it might work for ppc64le too. Not a blocking change, though, and I am no longer the maintainer here 🙂

File: recipe/meta.yaml (hunk 1 above)
Comment (bdice) [reply recipe/meta.yaml]: We need the linux constraint since this also has Windows builds.

Comment (leofang) [reply recipe/meta.yaml]: Right but `aarch64` evaluates to False on `win-32`/`win-64`.

Comment (leofang) [reply recipe/meta.yaml]: btw to keep for record, this is what I see with the CUDA 11 + aarch64 tarball:
```
$ for i in *.so; do echo $i; readelf -s $i | grep GLIBC | grep -v 'GLIBC_2.17'; done
libcudnn.so
libcudnn_adv_infer.so
libcudnn_adv_train.so
libcudnn_cnn_infer.so
   331: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND log2f@GLIBC_2.27 (12)
libcudnn_cnn_train.so
   118: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND powf@GLIBC_2.27 (8)
   250: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND log2f@GLIBC_2.27 (8)
   551: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND expf@GLIBC_2.27 (8)
libcudnn_ops_infer.so
libcudnn_ops_train.so
```

**Review by leofang — APPROVED** [review]

File: recipe/meta.yaml [hunk 2]
```diff
@@ -45,36 +47,46 @@ requirements:
...
     # to suppress ".so not found" errors
-    - libzlib  # [linux]
+    - libzlib       # [linux]
     - libzlib-wapi  # [win64]
   run:
+    - {{ pin_compatible("cuda-version", max_pin="x") }}
+    {% if cuda_major == "11" %}
     - cudatoolkit 11.*
+    {% else %}
+    - cuda-cudart-dev
```
Comment (kkraus14) [inline recipe/meta.yaml:64]: Why is the dev package required at runtime?

**Comment by kkraus14** [comment]
I also, would like to no longer be a recipe maintainer for this package. Is there any chance someone else would like to be added? cc @adibbley

**Comment by jakirkham** [comment]
We were going to propose adding some folks here. Would suggest:

- @adibbley
- @bdice
- @jakirkham
- @vyasr

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render

**Review by kkraus14 — APPROVED** [review]
Left one question about the dev package but otherwise lgtm. Thanks for taking over the maintenance of this package.

File: recipe/meta.yaml (hunk 2 above)
Comment (jakirkham) [reply recipe/meta.yaml:64]: Guess the issue is packages that depend on `cudnn` at runtime would now pull in these headers

Maybe we could put this in `run_constrained`? Admittedly that would serve a primarily documentation role

Alternatively we could rely on `cuda-nvcc_{{ target_platform }}` to pull in `cuda-cudart-dev`

Comment (bdice) [reply recipe/meta.yaml:64]: Sounds like dropping this is probably the consensus answer for now.

```suggestion
```

Comment (bdice) [reply recipe/meta.yaml]: Meant to add:

> Alternatively we could rely on cuda-nvcc_{{ target_platform }} to pull in cuda-cudart-dev

I agree this is sufficient, so I removed the dependency.

**Comment by jakirkham** [comment]
Thanks all! 🙏

Keith, could you please give this another look when you have a chance?

**Review by kkraus14 — APPROVED** [review]

**Comment by kkraus14** [comment]
Ideally this `cudnn` package should probably be split into a runtime and dev package, where the dev package then makes sense to depend on `cudart-dev`

**Comment by bdice** [comment]
> Ideally this `cudnn` package should probably be split into a runtime and dev package, where the dev package then makes sense to depend on `cudart-dev`

I think that is a good long-term plan. I'm hesitant to introduce that change right now, but the impact of that change can be surveyed and we can move in that direction over time. I'll file an issue.

**Comment by jakirkham** [comment]
Should add this same discussion came up offline previously and we decided to punt for now

**Comment by jakirkham** [comment]
Thanks all! 🙏

Going ahead and merging to get these packages out. Let's follow up on anything else separately 🙂

**Pull request merged**